openvpn/session/validator.go

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (C) 2018 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package session
+
+import (
+	"github.com/mysterium/node/identity"
+	"github.com/mysterium/node/session"
+	"errors"
+)
+
+// SignaturePrefix is used to prefix with each session string before calculating signature or extracting identity
+const SignaturePrefix = "MystVpnSessionId:"
+
+// Validator structure that keeps attributes needed Validator operations
+type Validator struct {
+	sessionManager *manager
+	identityExtractor identity.Extractor
+}
+
+// NewValidator return Validator instance
+func NewValidator(m *manager, extractor identity.Extractor) (*Validator) {
+	return &Validator{
+				sessionManager: m,
+				identityExtractor: extractor,
+			}
+}
+
+// Validate provides glue code for openvpn management interface to validate incoming client login request,
+// it expects session id as username, and session signature signed by client as password
+func (v *Validator) Validate(clientID int, sessionString, signatureString string) (bool, error) {
+		sessionID := session.SessionID(sessionString)
+		currentSession, found := v.sessionManager.FindUpdateSession(clientID, sessionID)
+		if !found {
+			return false, nil
+		}
+
+		signature := identity.SignatureBase64(signatureString)
+		extractedIdentity, err := v.identityExtractor.Extract([]byte(SignaturePrefix+sessionString), signature)
+		if err != nil {
+			return false, err
+		}
+		return currentSession.ConsumerID == extractedIdentity, nil
+}
+
+// Cleanup removes session from underlying session managers
+func (v *Validator) Cleanup(sessionString string) error {
```

Comment:
Cleanup is just callback from underlyging middleware, what caller has to do with returned error? Error handling (logging) can be performed here

Reply:
Cleanup is called as callback from middleware.handleClientEvent, error is handled there..

Reply:
That's my poiont. Why middleware which is lower level should handle cleanup errors? It doesn't know what to do with them anyway - except to log.

Reply:
I think its important to be able to test for error returned by Cleanup callback, which can be nicely tested for in tests as in TestCleanupReturnsErrorIfSessionNotExists()